### PR TITLE
fullnode server stub

### PR DESCRIPTION
**Pull request by mystenmark**
For now, we're simply implementing the read-only portion of the gateway API.

Questions:
  * Should we rename RpcServer to something more descriptive?
  * Should we split the RpcServer trait into read-only and write portions?

### CODE REVIEW

File: sui/src/bin/node.rs
Comment (bmwill) [inline sui/src/bin/node.rs:1]: I wonder if we can have a general purpose `sui-node` binary that can handle launching either a full-node or a validator instead of needing a different binary for every node type we want to have.

File: sui/src/sui_node.rs
Comment (bmwill) [inline sui/src/sui_node.rs:35]: Yeah I wonder if we couldn't, as a pre-step, split out the read and write apis into separate traits.

**Review by bmwill — COMMENTED** [review]
This generally looks good. Lets have @patrickkuo take a look since he owns the gateway service.

File: sui/src/bin/node.rs
Comment (mystenmark) [reply sui/src/bin/node.rs:1]: I suggested doing that earlier, but I believe @oxade had some objections. I think it would speed up docker builds significantly since we currently re-link several 99% identical binaries.

File: sui/src/sui_node.rs
Comment (mystenmark) [reply sui/src/sui_node.rs:35]: That's my inclination, but I wanted to get feedback first.

Comment (patrickkuo) [reply sui/src/sui_node.rs:35]: We can split the apis into separate traits, and instantiate them as different modules in the JSON-RPC server.

However it will be good if we can kick this can down the road if possible? (trying to prevent a rebase hell :( ), I am currently working on improving the API's response types, I am creating a set of new types for the Gateway instead of passing internal types (`Object`, `MoveObject`, `MoveType` etc) to downstream, as they doesn't work well with serde_json. Decoupling the object type at the GatewayAPI boundary gives us the flexibility to make the response more JSON friendly.

Comment (mystenmark) [reply sui/src/sui_node.rs:35]: Sure, I can wait until your PR is merged. I assume you need me to wait on this whole PR, not splitting RpcGateway?

Comment (patrickkuo) [reply sui/src/sui_node.rs:35]: Thanks!!
This PR is fine :) The changes in RpcGateway is quite straightforward.

**Review by bmwill — APPROVED** [review]

**Pull request merged**
